Shell/Builtin.cpp

```diff
@@ -0,0 +1,743 @@
+        }
+    }
+    if (!job) {
+        if (job_id == -1) {
+            printf("bg: no current job\n");
+        } else {
+            printf("bg: job with id %d not found\n", job_id);
+        }
+        return 1;
+    }
+
+    job->set_running_in_background(true);
+
+    dbg() << "Resuming " << job->pid() << " (" << job->cmd() << ")";
+    printf("Resuming job %llu - %s\n", job->job_id(), job->cmd().characters());
+
+    if (killpg(job->pgid(), SIGCONT) < 0) {
+        perror("killpg");
+        return 1;
+    }
+
+    return 0;
+}
+
+int Shell::builtin_cd(int argc, const char** argv)
+{
+    const char* arg_path = nullptr;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(arg_path, "Path to change to", "path", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    String new_path;
+
+    if (!arg_path) {
+        new_path = home;
+        if (cd_history.is_empty() || cd_history.last() != home)
+            cd_history.enqueue(home);
+    } else {
+        if (cd_history.is_empty() || cd_history.last() != arg_path)
+            cd_history.enqueue(arg_path);
+        if (strcmp(argv[1], "-") == 0) {
+            char* oldpwd = getenv("OLDPWD");
+            if (oldpwd == nullptr)
+                return 1;
+            new_path = oldpwd;
+        } else if (arg_path[0] == '/') {
+            new_path = argv[1];
+        } else {
+            StringBuilder builder;
+            builder.append(cwd);
+            builder.append('/');
+            builder.append(arg_path);
+            new_path = builder.to_string();
+        }
+    }
+
+    LexicalPath lexical_path(new_path);
+    if (!lexical_path.is_valid()) {
+        printf("LexicalPath failed to canonicalize '%s'\n", new_path.characters());
+        return 1;
+    }
+    const char* path = lexical_path.string().characters();
+
+    struct stat st;
+    int rc = stat(path, &st);
+    if (rc < 0) {
+        printf("stat(%s) failed: %s\n", path, strerror(errno));
+        return 1;
+    }
+    if (!S_ISDIR(st.st_mode)) {
+        printf("Not a directory: %s\n", path);
+        return 1;
+    }
+    rc = chdir(path);
+    if (rc < 0) {
+        printf("chdir(%s) failed: %s\n", path, strerror(errno));
+        return 1;
+    }
+    setenv("OLDPWD", cwd.characters(), 1);
+    cwd = lexical_path.string();
+    setenv("PWD", cwd.characters(), 1);
+    return 0;
+}
+
+int Shell::builtin_cdh(int argc, const char** argv)
+{
+    int index = -1;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(index, "Index of the cd history entry (leave out for a list)", "index", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (index == -1) {
+        if (cd_history.size() == 0) {
+            printf("cdh: no history available\n");
+            return 0;
+        }
+
+        for (int i = cd_history.size() - 1; i >= 0; --i)
```

Comment:
```suggestion
        for (ssize_t i = cd_history.size() - 1; i >= 0; --i)
```

Reply:
It seems that's intentional.
the value goes to `-1` to signal the end.
I'll change the logic along with this.

Reply:
Take responsibility:
`cdh 1` ->
```
[Shell(24:24)]: NP(error) fault at invalid address V0x00000001
[Shell(24:24)]: Unrecoverable page fault, read from address V0x00000001
[Shell(24:24)]: Note: Address V0x00000001 looks like a possible nullptr dereference
[Shell(24:24)]: CRASH: Page Fault. Ring 3.
[Shell(24:24)]: exception code: 0004 (isr: 0000
[Shell(24:24)]:   pc=001b:0807bbe3 flags=0246
[Shell(24:24)]:  stk=0023:025020f8
[Shell(24:24)]:   ds=0023 es=0023 fs=0023 gs=07d3
[Shell(24:24)]: eax=00000000 ebx=02502198 ecx=00000001 edx=00000001
[Shell(24:24)]: ebp=025020f8 esp=c371ffe8 esi=02502154 edi=025021ac
[Shell(24:24)]: cr0=8001001b cr2=00000001 cr3=01c24000 cr4=00300ee4
[Shell(24:24)]: code: 80 3c 02 00 74 03 40 eb
[Shell(24:24)]: 0x0807bbe3  strlen +8

[Shell(24:24)]: 0xc015b79c  Kernel::Process::crash(int, unsigned int, bool) +906
[Shell(24:24)]: 0xc011fb01  illegal_instruction_handler +0
[Shell(24:24)]: 0xc0120dff  page_fault_handler +836
[Shell(24:24)]: 0xc011e815  page_fault_asm_entry +22
[Shell(24:24)]: 0x0806fee1  (anonymous namespace)::OptionParser::getopt() +181
[Shell(24:24)]: 0x0807053c  getopt_long +192
[Shell(24:24)]: 0x0805f8b4  Core::ArgsParser::parse(int, char**, bool) +328
[Shell(24:24)]: 0x0805340c  Shell::builtin_cd(int, char const**) +72
[Shell(24:24)]: 0x0805383b  Shell::builtin_cdh(int, char const**) +277
[Shell(24:24)]: 0x08055173  Shell::run_builtin(int, char const**, int&) +177
[Shell(24:24)]: 0x0805ce25  Shell::run_command(AST::Command&) +1099
[Shell(24:24)]: 0x080501c5  auto AST::Execute::run(AK::RefPtr<Shell>)::{lambda(auto:1&)#1}::operator()<AK::Vector<AST::Command, 0ul> >(AK::Vector<AST::Command, 0ul>&) const +89
[Shell(24:24)]: 0x08050df8  AST::Execute::run(AK::RefPtr<Shell>) +2520
[Shell(24:24)]: 0x0805c5f4  Shell::run_command(AK::StringView const&) +214
[Shell(24:24)]: 0x0805c8fa  Shell::read_single_line() +272
[Shell(24:24)]: 0x0805c96f  Shell::custom_event(Core::CustomEvent&) +29
[Shell(24:24)]: 0x080664b0  Core::Object::dispatch_event(Core::Event&, Core::Object*) +74
[Shell(24:24)]: 0x080611c3  Core::EventLoop::pump(Core::EventLoop::WaitMode) +307
[Shell(24:24)]: 0x08061341  Core::EventLoop::exec() +43
[Shell(24:24)]: 0x08048933  main +1355
[Shell(24:24)]: 0x08048c2e  _start +94
```

Reply:
You lucked out, it was `cdh` not putting the terminating null in argv.